### PR TITLE
bpo-1294959: Add --with-platlibdir to configure

### PR DESCRIPTION
**Pull request by vstinner**
Add --with-platlibdir option to configure to choose where
platform-specific and standard library files are installed
(default: 'lib'). It is used on Fedora and SuSE to use /usr/lib64
instead of /usr/lib on 64-bit systems.

Add sys.platlibdir attribute.

Co-Authored-By: Jan Matějek <jmatejek@suse.com>
Co-Authored-By: Matěj Cepl <mcepl@cepl.eu>
Co-Authored-By: Charalampos Stratakis <cstratak@redhat.com>

https://bugs.python.org/issue1294959

### CODE REVIEW

**Comment by vstinner** [comment]
Fedora is maintaining a downstream patch for years in the different python packages: https://src.fedoraproject.org/rpms/python3/blob/master/f/00102-lib64.patch

This PR is not this patch, it's a rewrite based on the Fedora: we made the library name configurable.

This PR doesn't change the default which remains "lib". It only adds an opt-in option for Linux vendors like Fedora and OpenSUSE who already change this directory. I'm not sure about Debian/Ubuntu (I have to check).

Previous attempts to implement https://bugs.python.org/issue1294959:

* PR #8068 by @mcepl (2018)
* PR #11755 by @stratakis (2019)

There is also https://bugs.python.org/issue34032 which should be marked as a duplicate.

**Comment by vstinner** [comment]
test_embed is failing. I have to investigate to understand why, so I marked the PR as a WIP.

Note: I excluded the setup.py change of the Fedora change from this PR, it doesn't seem to be **directly** related to this PR.

**Comment by vstinner** [comment]
> Note: I excluded the setup.py change of the Fedora change from this PR, it doesn't seem to be directly related to this PR.

Hum, once this PR wil lbe merged, I will work on a separated PR for setup.py.

File: Lib/site.py
Comment (vstinner) [inline Lib/site.py]:
```suggestion
                                            "python%d.%d" % sys.version_info[:2],
```

**Comment by vstinner** [comment]
I compared this PR with PR #3698 written by Jan Matějek for OpenSUSE. They look **very similar**. The main differences are that:

* (1) this PR also updates the path for the stdlib. For example, on Fedora, the os module is stored in `/usr/lib64/python3.7/os.py` (**lib64**), not in ``/usr/lib/python3.7/os.py`` (**lib**).
* (2) this PR also updates the paths for the user paths. Example on Fedora:

```
$ python3
Python 3.7.6 (default, Jan 30 2020, 09:44:41) 
>>> import sysconfig, pprint
>>> pprint.pprint(sysconfig.get_paths('posix_user'))
{'data': '/home/vstinner/.local',
 'include': '/home/vstinner/.local/include/python3.7',
 'platlib': '/home/vstinner/.local/lib64/python3.7/site-packages',  # <=== HERE
 'platstdlib': '/home/vstinner/.local/lib64/python3.7',  # <=== HERE
 'purelib': '/home/vstinner/.local/lib/python3.7/site-packages',
 'scripts': '/home/vstinner/.local/bin',
 'stdlib': '/home/vstinner/.local/lib64/python3.7'}  # <=== HERE
```

I don't know how it's used. In practice, I only have ``~/.local/lib/python3.7/`` in my HOME directory and it does contain a few C extensions:

```
/home/vstinner/.local/lib/python3.7/site-packages/lazy_object_proxy/cext.cpython-37m-x86_64-linux-gnu.so
/home/vstinner/.local/lib/python3.7/site-packages/typed_ast/_ast27.cpython-37m-x86_64-linux-gnu.so
/home/vstinner/.local/lib/python3.7/site-packages/typed_ast/_ast3.cpython-37m-x86_64-linux-gnu.so
```

---

All differences:

* This PR adds **sys.python_libdir** variable, Matějek's PR uses get_config_var("platsubdir") instead
* Lib/distutils/command/install.py: this PR also changes **unix_home['platlib']**
* Lib/sysconfig.py: this PR also changes **posix_user scheme**
* Lib/distutils/sysconfig.py: this PR uses python_libdir if **standard_lib is true** or plat_specific is true, Matějek's PR only uses platsubdir for plat_specific is true. Side effect: this PR has to update pydoc as well for the stdlib path.
* test_site.py: this PR checks also dirs[1], Matějek's PR only checks dirs[0]
* Matějek's PR also changes Lib/test/test_sysconfig.py for test on posix_user
* Matějek's PR also updates Lib/trace.py if ignore dirs is used: it looks like a bug in this PR
* Makefile.pre.in: this PR uses variable "PYTHON_LIBDIR", Matějek's PR uses "platsubdir" variable
* Matějek's PR adds two long Changelog entries :-)
* Modules/getpath.c: this PR updates also <python_dirlib> / "lib-dynload" path, Matějek's PR only updates "lib/python" path.
* configure.ac: this PR adds --with-python-libdir=lib, Matějek's PR adds --with-custom-platsubdir=lib

**Comment by vstinner** [comment]
> (1) this PR also updates the path for the stdlib. For example, on Fedora, the os module is stored in /usr/lib64/python3.7/os.py (lib64), not in /usr/lib/python3.7/os.py (lib).

I guess that the intent was to be able to co-install i686 (32-bit) and x86-64 (64-bit) flavors of Python 3 without having conflicts on paths.

For example, python3-libs.i686 installs os.py into /usr/**lib**/python3.7/os.py, whereas python3-libs.x86_64 installs os.py into /usr/**lib64**/python3.7/os.py.

On Fedora 31, I can co-install python3.i686 and python3-libs.i686, whereas python3.x86_64 and python3-libs.x86_64 are already installed. But I failed to run the 32-bit version of /usr/bin/python3.7. This file is provided by python3.i686 and python3.x86_64, but it seems like the 64-bit flavor won.

**Comment by vstinner** [comment]
@stratakis's PR #11755 is very similar to @matejcik's PR #3698. He even explained it: "This PR is based on #3698 rebased on top of master and the NEWS entries tidied up a bit. I haven't tested it yet as of recently thus the WIP in the title."

**Comment by vstinner** [comment]
On SUSE, the stdlib is also in /usr/lib64. Example: /usr/lib64/python3.7/os.py

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@RandyMcMillan

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by vstinner** [comment]
Hum, I messed up and got unreleated commits in this PR. I rebased it on top to master to exclude these unrelated commits.

**Comment by vstinner** [comment]
@mcepl: Would it be possible for you to replace SuSE downstream patches and check that you still get the same paths? You can use  lib64_tests.py script attached to https://bugs.python.org/issue1294959 to check that you get the same paths before/after. You might adapt this script for SuSE. I wrote the script for Fedora.

File: Doc/library/sys.rst
Comment (brettcannon) [inline Doc/library/sys.rst]: Why not `libdir`? The `python` bit seems a bit redundant.

File: Doc/library/sys.rst
Comment (brettcannon) [inline Doc/library/sys.rst]: Maybe specify this is where `site-packages` will end up (e.g. `f"{sys.libdir}/python{'.'.join([sys.version_info[:2]}/site-packages"`)? Otherwise I had to look at the diff in the Makefile to understand what this directory represented.

File: Lib/distutils/sysconfig.py
Comment (brettcannon) [inline Lib/distutils/sysconfig.py]:
```suggestion
            # module distribution) or standard Python library modules.
```

File: Doc/library/sys.rst
Comment (brettcannon) [reply Doc/library/sys.rst]: And what is the value under Windows?

File: Lib/distutils/sysconfig.py
Comment (brettcannon) [inline Lib/distutils/sysconfig.py:155]: If this is the default then why bother testing for the value and simply use `sys.python_libdir` directly?

File: Lib/site.py
Comment (brettcannon) [inline Lib/site.py]:
```suggestion
            libdirs = ["lib"]
            if sys.python_libdir != "lib":
                libdirs.insert(0, sys.python_libdir)
            for libdir in libdirs:
                dir = os.path.join(prefix, libdir,
                                        "python%d.%d" % sys.version_info[:2],
                                        "site-packages")
            sitepackages.append(dir)
```

Same for the `else` block to avoid the copy-and-paste of the path creation.

**Review by brettcannon — DISMISSED** [review]
Probably need to expand the docs a bit as well as a suggestion on how to simplify some code.

**Comment by brettcannon** [comment]
@mcepl any chance you can check if this will work for you as Victor asked?

File: Doc/library/sys.rst
Comment (stratakis) [reply Doc/library/sys.rst]: The reason for that is that `libdir` is an autoconf variable: https://www.gnu.org/prep/standards/html_node/Directory-Variables.html

Comment (vstinner) [reply Doc/library/sys.rst]: I hesitate between python_libdir and libdir. I'm fine with sys.libdir: sys.prefix, sys.executable, etc. don't have "python_" prefix.

But yes, the "python" prefix is to avoid conflict with autotools standard --libdir option of ./configure script: this option has a different meaning.

**Comment by vstinner** [comment]
@brettcannon: sys.python_libdir is equal to "lib" on Windows. It's equal to all platforms, except on Fedora and OpenSUSE in short.

Example on unmodified Python on Windows:

```
>>> for path in sys.path: print(path)
... 

C:\Users\vstinner\AppData\Local\Programs\Python\Python39\python39.zip
C:\Users\vstinner\AppData\Local\Programs\Python\Python39\DLLs
C:\Users\vstinner\AppData\Local\Programs\Python\Python39\lib
C:\Users\vstinner\AppData\Local\Programs\Python\Python39
C:\Users\vstinner\AppData\Roaming\Python\Python39\site-packages
C:\Users\vstinner\AppData\Local\Programs\Python\Python39\lib\site-packages   
```

See the ``lib\site-packages`` path in ``C:\Users\vstinner\(...)``: it includes the ``lib`` directory.

**Comment by vstinner** [comment]
@brettcannon: "@mcepl any chance you can check if this will work for you as Victor asked?"

I'm in touch with him. The problem is that currently, OpenSUSE uses Python 3.8, whereas this change is written for Python 3.9.

I may try to backport the change to 3.8, just to provide a patch to @mcepl, so he can test on OpenSUSE, to make sure that it works as expected.

**Comment by vstinner** [comment]
> I may try to backport the change to 3.8, just to provide a patch to @mcepl, so he can test on OpenSUSE, to make sure that it works as expected.

Let me clarify: I only offer to provide a temporary patch, not backport this new feature to 3.8 ;-)

**Comment by mcepl** [comment]
> Let me clarify: I only offer to provide a temporary patch, not backport this new feature to 3.8 ;-)

That is not required. After we finally hopefully (I hate [salt](https://community.saltstack.com/)!) get Python 3.8 into Factory, I will start building 3.9 in our development project. So, I have [started](https://build.opensuse.org/package/show/home:mcepl:branches:devel:languages:python:Factory/python3) in my home project, and it goes slowly.

**Comment by mcepl** [comment]
Even with (generated, so too long) configure command:
```
./configure --host=x86_64-suse-linux-gnu \
    --build=x86_64-suse-linux-gnu --program-prefix= \
    --disable-dependency-tracking --prefix=/usr --exec-prefix=/usr \
    --bindir=/usr/bin --sbindir=/usr/sbin --sysconfdir=/etc \
    --datadir=/usr/share --includedir=/usr/include \
    --libdir=/usr/lib64 --libexecdir=/usr/lib --localstatedir=/var \
    --sharedstatedir=/var/lib --mandir=/usr/share/man \
    --infodir=/usr/share/info --disable-dependency-tracking \
    --docdir=/usr/share/doc/packages/python \
    --with-python-libdir=lib64 --enable-ipv6 --enable-shared \
    --with-ensurepip=no --with-system-ffi --with-system-expat \
    --with-lto --enable-optimizations \
    --enable-loadable-sqlite-extensions
```
I get ``/usr/python3.9/config-3.9-x86_64-linux-gnu/`` directory for all files there, except for ``libpython3.9.so``, which goes correctly to ``/usr/lib64/python3.9/config-3.9-x86_64-linux-gnu/libpython3.9.so``.

When checking the generated ``Makefile`` I see the line 1573:
```Makefile
LIBPL=          $(prefix)//python3.9/config-$(VERSION)$(ABIFLAGS)-x86_64-linux-gnu
```
which seems to me obviously wrong. I would think it should be
```Makefile
LIBPL=          $(LIBDIR)/python3.9/config-$(VERSION)$(ABIFLAGS)-x86_64-linux-gnu
```

File: Doc/library/sys.rst
Comment (brettcannon) [reply Doc/library/sys.rst]: @stratakis  sure, but the `sys` module isn't autoconf-specific 😉 .

**Comment by mcepl** [comment]
@vstinner OK, aside from the above problem with config-* subdir and bpo#39855, Python 3.9.0a3 builds correctly and all tests pass. +1 from me.

**Comment by vstinner** [comment]
> LIBPL=          $(prefix)//python3.9/config-$(VERSION)$(ABIFLAGS)-x86_64-linux-gnu

Oops. That's a stupid typo, it's now fixed. Well spotted, thanks!

File: Doc/library/sys.rst
Comment (merwok) [reply Doc/library/sys.rst]: I saw `libdir_name` in a previous commit, that sounds like a clear name to me!

File: Lib/distutils/sysconfig.py
Comment (vstinner) [reply Lib/distutils/sysconfig.py:155]: When we want lib64, this code path must continue to use "lib". Example of Fedora:

```
>>> get_python_lib()
'/usr/lib/python3.7/site-packages'
>>> get_python_lib(standard_lib=1)
'/usr/lib64/python3.7'
```

/usr/lib/python3.7/site-packages contains all third party modules. For example, pip is installed there by the python3-pip RPM package.

whereas the stdlib lives in /usr/lib64:

```
$ python3
Python 3.7.6 (default, Jan 30 2020, 09:44:41) 
[GCC 9.2.1 20190827 (Red Hat 9.2.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os
<module 'os' from '/usr/lib64/python3.7/os.py'>
```

The idea is to be able to co-install Python in 32-bit and 64-bit mode: each Python gets its own standard library directory (``/usr/lib/python3.7/os.py`` for 32-bit, ``/usr/lib64/python3.7/os.py`` for 64-bit), whereas third party modules are shared.

File: Lib/site.py
Comment (vstinner) [reply Lib/site.py]: Right. I merged duplicated code based on your proposal, thanks.

**Comment by mcepl** [comment]
Output of ``lib64_tests.py`` reports one failure:
```
[  507s] ======================================================================
[  507s] FAIL: test_sys (__main__.Lib64Tests)
[  507s] ----------------------------------------------------------------------
[  507s] Traceback (most recent call last):
[  507s]   File "/home/abuild/rpmbuild/SOURCES/lib64_tests.py", line 47, in test_sys
[  507s]     self.assertEqual(paths[0], f'/usr/lib64/python3{MINOR}.zip')
[  507s] AssertionError: '/home/abuild/rpmbuild/SOURCES' != '/usr/lib64/python39.zip'
[  507s] - /home/abuild/rpmbuild/SOURCES
[  507s] + /usr/lib64/python39.zip
[  507s]
```

**Comment by vstinner** [comment]
Ok ok, we reached the point where the most important question became... the **option name** :-D There are several choices:

* ``./configure --libdir-name=lib64`` and ``sys.libdir_name``: @merwok likes it, **me** too!
* ``./configure --libsubdir=lib64`` and ``sys.libsubdir``
* ``./configure --python-libdir=lib64`` and ``sys.python_libdir``: let me be plain honest, I don't like my own proposal (current choice of this PR)...
* ``./configure --with-custom-platsubdir``: SuSE downstream patches
* ``./configure --with-custom-platlibdir``: proposed in https://bugs.python.org/issue34032 (variant of the SuSE patch)

@brettcannon @encukou @stratakis: What's your call on this very important question? :-D

Not considered:

* ``sys.implementation._lib`` or ``sys.implementation.platlibdir``: Matthias Klose disagreed with using sys.implementation; https://bugs.python.org/issue1294959#msg202366
* ``./configure --libdir=lib64`` and ``sys.libdir``: I don't think that anyone considered this option seriously, --libdir is a standard configure/autotools option which has a different meaning.

---

About the need for adding a new sys attribute.  SuSE doesn't add a sys attribute, instead it imports sysconfig in site which has a negative impact on Python startup performance. I chose to add an attribute to sys to avoid that.

**Comment by vstinner** [comment]
@mcepl: "Output of lib64_tests.py reports one failure: (...) AssertionError: '/home/abuild/rpmbuild/SOURCES' != '/usr/lib64/python39.zip'"

Well, it's really hard to test sys.path. You can ignore this error. It seems like something injected SOURCES in sys.path which wasn't expected by the test. At least, other tests passed which is great!

**Comment by encukou** [comment]
Since you explicitly asked for bikeshedding:
* `libdir_name` looks like "the name of `libdir`"
* `libsubdir` looks like "a subdirectory of `libdir`"
* `python-libdir` looks like "`libdir`, but specific to Python"
* `platsubdir` looks like "a subdirectory of `platdir`"
* `platlibdir` looks like "a platform-specific variant of `libdir`"

I'd go for `platlibdir`. But I have no strong preference.

**Comment by stratakis** [comment]
I concur with Petr. I really like the `--with-custom-platlibdir` and as a second choice I would go for `--libdir-name`

**Comment by brettcannon** [comment]
👍 to Petr's suggestion

**Comment by vstinner** [comment]
I updated the PR to use ``./configure --with-platlibdir`` and ``sys.platlibdir`` names.

I completed the list of authors::

```
    Co-Authored-By: Jan Matějek <jmatejek@suse.com>
    Co-Authored-By: Matěj Cepl <mcepl@cepl.eu>
    Co-Authored-By: Charalampos Stratakis <cstratak@redhat.com>
```

https://bugs.python.org/issue1294959 has a long history, and these 3 developers wrote earlier patche and pull requests. We all collaborated somehow.

> @stratakis: "I concur with Petr. I really like the --with-custom-platlibdir (...)"

I dislike the ``custom-`` prefix. There is no other configure option which such prefix.

**Comment by stratakis** [comment]
From my perspective the PR looks good to me.

**Comment by vstinner** [comment]
> I updated the PR to use ./configure --with-platlibdir and sys.platlibdir names.

Changes:

* PR rebased on master
* Commits squashed to rewrite the commit message
* sys.platlibdir documentation completed to give more concrete examples of paths
* Fix a memory leak in sysmodule.c if PyDict_SetItem() fails: use SET_SYS_FROM_STRING, rather than SET_SYS_FROM_STRING_BORROW
* site: I also factorized duplicated code for the os.sep != "/" (Windows) code path

**Comment by mcepl** [comment]
> I updated the PR to use `./configure --with-platlibdir` and `sys.platlibdir` names.

+1 from me. It [builds](https://build.opensuse.org/package/show/home:mcepl:branches:devel:languages:python:Factory/python3) without any problem.

**Comment by vstinner** [comment]
I updated the PR to complete/adjust the documentation.

It works as expected. I tested my PR with:

```
git clean -fdx
./configure --prefix /usr --with-platlibdir=lib64 CFLAGS="-O0"
make
curl -O https://bugs.python.org/file48966/lib64_tests-3.py
./python lib64_tests-3.py 
```

Output:

```
.....
WARNING: Python is run from its source directory!
PR 18381 (sys.platlibdir)? True

----------------------------------------------------------------------
Ran 5 tests in 0.032s

OK
```

**Comment by vstinner** [comment]
I created https://src.fedoraproject.org/rpms/python39/pull-request/31 to test this PR on the python39 RPM package for Fedora 31 (Stable).

**Comment by vstinner** [comment]
> I created https://src.fedoraproject.org/rpms/python39/pull-request/31 to test this PR on the python39 RPM package for Fedora 31 (Stable).

The scratch build (temporary package for testing) completed successfully on x86-64: the patch applies and built Python uses /usr/lib64 as expected.

I tested manually lib64_tests-3.py of https://bugs.python.org/issue1294959: paths are unchanged with this PR. It means that https://github.com/python/cpython/pull/18381 works as expected.

**Pull request merged**
